### PR TITLE
qos_hex: Fix edge cases with short inputs to decode

### PR DESCRIPTION
**Pull request by emostov**
This try's to address panics found from fuzzing users of `qos_hex`. We were getting issues of indexing out of bounds in this line:

https://github.com/tkhq/qos/blob/780176eec3d4ab52a33b638784d92935ddcc680f/src/qos_hex/src/lib.rs#L49

because slicing into a str with a length less then 2 panics.

This fixes errors like:

```
(lldb) r ../fuzzing/out_evm_parser.round2/default/crashes/id:000013,sig:06,src:000265+000210,time:9731995,execs:118583205,op:splice,rep:4 
Process 60555 launched: 'qos-main/src/target/debug/fuzz_evm_parser_replay' (x86_64)
thread 'main' panicked at 'byte index 2 is out of bounds of `
`', qos_hex/src/lib.rs:49:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Process 60555 exited with status = 101 (0x00000065)
```

### CODE REVIEW

File: src/qos_hex/src/lib.rs [hunk 1]
```diff
@@ -45,17 +47,30 @@ impl From<ParseIntError> for HexError {
...
-	let is_even_len = str_byte_len % 2 == 0;
-	let is_lt_max_len = str_byte_len < STR_MAX_LENGTH;
+pub fn decode(raw_s: &str) -> Result<Vec<u8>, HexError> {
+	let sanitized_s = {
+		let raw_s_byte_len = raw_s.len();
+		if raw_s_byte_len == 0 {
+			// We can do an explicit exit early
+			return Ok(Vec::new());
+		} else if raw_s_byte_len == 1 {
+			return Err(HexError::LengthOne);
```
Comment (emostov) [inline src/qos_hex/src/lib.rs]: We could use the `OddLength` error, but I thought being more specific could be helpful.

File: src/qos_hex/src/lib.rs [hunk 2]
```diff
@@ -45,17 +47,30 @@ impl From<ParseIntError> for HexError {
...
-	let str_byte_len = s.len();
-
-	let is_even_len = str_byte_len % 2 == 0;
-	let is_lt_max_len = str_byte_len < STR_MAX_LENGTH;
+pub fn decode(raw_s: &str) -> Result<Vec<u8>, HexError> {
+	let sanitized_s = {
+		let raw_s_byte_len = raw_s.len();
+		if raw_s_byte_len == 0 {
+			// We can do an explicit exit early
+			return Ok(Vec::new());
```
Comment (emostov) [inline src/qos_hex/src/lib.rs]: We could just do a soft return of `raw_s` and then go through the rest of the function, but I thought the explicit early exit was easier to follow

File: src/qos_hex/src/lib.rs [hunk 3]
```diff
@@ -82,6 +97,24 @@ pub fn encode(bytes: &[u8]) -> String {
...
+	#[test]
+	fn decode_correctly_errors_with_len_one() {
+		let encoded = "a";
+		let res = decode(encoded);
+		assert_eq!(res, Err(HexError::LengthOne));
+
+		let encoded = " ";
+		let res = decode(encoded);
+		assert_eq!(res, Err(HexError::LengthOne));
+	}
```
Comment (emostov) [inline src/qos_hex/src/lib.rs:114]: The two above tests check for the edge cases encountered while fuzzing

File: src/qos_hex/src/lib.rs [hunk 4]
```diff
@@ -94,16 +127,22 @@ mod test {
 		assert_eq!(encode(&decoded), encoded);
```
Comment (emostov) [inline src/qos_hex/src/lib.rs:125]: below diff is just breaking up tests so it they are easier to follow and catch exactly where things break

File: src/qos_hex/src/lib.rs [hunk 5]
```diff
@@ -45,17 +47,30 @@ impl From<ParseIntError> for HexError {
...
+			// We can do an explicit exit early
+			return Ok(Vec::new());
+		} else if raw_s_byte_len == 1 {
+			return Err(HexError::LengthOne);
+		} else if &raw_s[..2] == "0x" {
+			// we know the s is at least len 2
+			&raw_s[2..]
+		} else {
+			raw_s
+		}
```
Comment (r-n-o) [inline src/qos_hex/src/lib.rs]: To make the code more readable, you can keep the top-level conditional based on length only. That way there's no need for a comment "we know the s is at least len 2". It's obvious from the construction:
```rust
if raw_s_byte_len == 0 {
  // handle empty case
} else if raw_s_byte_len == 1 {
  // handle single char case
} else if raw_s_byte_len >=2 {
  if &raw_s[..2] == "0x" {...}
  else { raw_s }
} else {
  // negative length? Something's gone massively wrong...
}
```

Comment (emostov) [reply src/qos_hex/src/lib.rs]: we have a usize, so can't be negative. But would be nice to somehow leverage the compiler to assert that if the length isn't 0 or 1 we know that it is always at least 2

Comment (r-n-o) [reply src/qos_hex/src/lib.rs]: Maybe with a `match` statement? Would the following work?
```rust
match raw_s_byte_len {
    0 => ...(empty case)
    1 => ...(single char case)
    _ => ...(case with >= 2)
}

Comment (emostov) [reply src/qos_hex/src/lib.rs]: we had the same thought!

https://github.com/tkhq/qos/pull/193/commits/e530c13928a9354ec5d064983b1b863d83f54f58
